cedar-policy/src/tests.rs

```diff
@@ -4862,7 +4915,394 @@ mod policy_set_est_tests {
+            Context::empty(),
+            Some(&schema),
+        )
+        .unwrap_err();
+        expect_err(
+            "",
+            &Report::new(err),
+            &ExpectedErrorMessageBuilder::error(
+                "principal type `Undeclared` is not declared in the schema",
+            )
+            .build(),
+        );
+    }
+
+    #[test]
+    fn undeclared_resource_type() {
+        let schema = schema();
+        let err = Request::new(
+            Some(EntityUid::from_strs("Principal", "principal")),
+            Some(EntityUid::from_strs("Action", "action")),
+            Some(EntityUid::from_strs("Undeclared", "resource")),
+            Context::empty(),
+            Some(&schema),
+        )
+        .unwrap_err();
+        expect_err(
+            "",
+            &Report::new(err),
+            &ExpectedErrorMessageBuilder::error(
+                "resource type `Undeclared` is not declared in the schema",
+            )
+            .build(),
+        );
+    }
+
+    #[test]
+    fn invalid_principal_type() {
+        let schema = schema();
+        let err = Request::new(
+            Some(EntityUid::from_strs("Resource", "principal")),
+            Some(EntityUid::from_strs("Action", "action")),
+            Some(EntityUid::from_strs("Resource", "resource")),
+            Context::empty(),
+            Some(&schema),
+        )
+        .unwrap_err();
+        expect_err(
+            "",
+            &Report::new(err),
+            &ExpectedErrorMessageBuilder::error(
+                r#"principal type `Resource` is not valid for `Action::"action"`"#,
+            )
+            .build(),
+        );
+    }
+
+    #[test]
+    fn invalid_resource_type() {
+        let schema = schema();
+        let err = Request::new(
+            Some(EntityUid::from_strs("Principal", "principal")),
+            Some(EntityUid::from_strs("Action", "action")),
+            Some(EntityUid::from_strs("Principal", "resource")),
+            Context::empty(),
+            Some(&schema),
+        )
+        .unwrap_err();
+        expect_err(
+            "",
+            &Report::new(err),
+            &ExpectedErrorMessageBuilder::error(
+                r#"resource type `Principal` is not valid for `Action::"action"`"#,
+            )
+            .build(),
+        );
+    }
+
+    #[test]
+    fn invalid_context() {
+        let schema = schema();
+        let err = Request::new(
+            Some(EntityUid::from_strs("Principal", "principal")),
+            Some(EntityUid::from_strs("Action", "action")),
+            Some(EntityUid::from_strs("Resource", "resource")),
+            Context::empty(),
+            Some(&schema),
+        )
+        .unwrap_err();
+        expect_err(
+            "",
+            &Report::new(err),
+            &ExpectedErrorMessageBuilder::error(
+                r#"context `<first-class record with 0 fields>` is not valid for `Action::"action"`"#,
```

Comment:
it would probably be more helpful to say
```
context `{}` is not valid for `Action::"action"`
```

Reply:
Punting to future work 🙂 #998